File: shared/clawhiveJwt.ts
/**
 * ClawHive JWT（HS256）本地验签 —— 与 Manage-platform auth.create_access_token 对齐。
 * claims：sub(=username=userId) / role / tenant_id / exp
 * 无 jose 依赖：Node crypto 验签。
 */

import { createHmac, timingSafeEqual } from 'node:crypto'

export type ClawhiveJwtClaims = {
  sub: string
  role?: string
  tenant_id?: string
  exp?: number
  iat?: number
}

export type ClawhiveUser = {
  userId: string
  username: string
  role: string
  tenantId: string
}

function b64urlToBuf(input: string): Buffer {
  const s = String(input || '').replace(/-/g, '+').replace(/_/g, '/')
  const pad = s.length % 4 === 0 ? '' : '='.repeat(4 - (s.length % 4))
  return Buffer.from(s + pad, 'base64')
}

function b64urlJson(input: string): Record<string, unknown> {
  const raw = b64urlToBuf(input).toString('utf8')
  const o = JSON.parse(raw)
  return o && typeof o === 'object' ? (o as Record<string, unknown>) : {}
}

export function resolveJwtSecret(env: NodeJS.ProcessEnv = process.env): string {
  return String(env.JWT_SECRET || env.CLAWHIVE_JWT_SECRET || '').trim()
}

export function isAgentBrowserAuthEnabled(env: NodeJS.ProcessEnv = process.env): boolean {
  const v = String(env.AGENT_BROWSER_AUTH ?? env.MANAGER_USER_AUTH ?? '').trim()
  if (v === '0' || v.toLowerCase() === 'false') return false
  if (v === '1' || v.toLowerCase() === 'true') return true
  // Docker / 配了 JWT_SECRET 时默认开
  return Boolean(resolveJwtSecret(env))
}

export function extractBearer(authorization?: string | null): string {
  const m = /^Bearer\s+(.+)$/i.exec(String(authorization || '').trim())
  return m ? String(m[1] || '').trim() : ''
}

export function verifyClawhiveJwt(
  token: string,
  env: NodeJS.ProcessEnv = process.env
): ClawhiveJwtClaims {
  const secret = resolveJwtSecret(env)
  if (!secret) throw new Error('JWT_SECRET_not_configured')
  const parts = String(token || '').trim().split('.')
  if (parts.length !== 3) throw new Error('jwt_malformed')
  const [h, p, s] = parts
  const data = `${h}.${p}`
  const expected = createHmac('sha256', secret).update(data).digest()
  let got: Buffer
  try {
    got = b64urlToBuf(s)
  } catch {
    throw new Error('jwt_sig_invalid')
  }
  if (got.length !== expected.length || !timingSafeEqual(got, expected)) {
    throw new Error('jwt_sig_invalid')
  }
  const header = b64urlJson(h)
  const alg = String(header.alg || '')
  if (alg && alg !== 'HS256') throw new Error('jwt_alg_unsupported')
  const payload = b64urlJson(p)
  const sub = String(payload.sub || '').trim()
  if (!sub) throw new Error('jwt_missing_sub')
  const exp = Number(payload.exp)
  if (Number.isFinite(exp) && exp > 0 && Date.now() / 1000 >= exp) {
    throw new Error('jwt_expired')
  }
  return {
    sub,
    role: payload.role != null ? String(payload.role) : undefined,
    tenant_id: payload.tenant_id != null ? String(payload.tenant_id) : undefined,
    exp: Number.isFinite(exp) ? exp : undefined,
    iat: Number.isFinite(Number(payload.iat)) ? Number(payload.iat) : undefined
  }
}

export function resolveBrowserUser(token: string, env: NodeJS.ProcessEnv = process.env): ClawhiveUser {
  const claims = verifyClawhiveJwt(token, env)
  return {
    userId: claims.sub,
    username: claims.sub,
    role: String(claims.role || 'viewer'),
    tenantId: String(claims.tenant_id || 'default')
  }
}

export function tryResolveBrowserUser(
  token: string | undefined | null,
  env: NodeJS.ProcessEnv = process.env
): ClawhiveUser | null {
  const t = String(token || '').trim()
  if (!t) return null
  try {
    return resolveBrowserUser(t, env)
  } catch {
    return null
  }
}

File: shared/clawhive_jwt.py
"""ClawHive JWT（HS256）本地验签 —— 与 Manage-platform auth.create_access_token 对齐。"""

from __future__ import annotations

import base64
import hashlib
import hmac
import json
import os
import time
from typing import Any


class ClawhiveAuthError(Exception):
    def __init__(self, code: str, message: str = "") -> None:
        self.code = code
        super().__init__(message or code)


def resolve_jwt_secret(env: dict[str, str] | None = None) -> str:
    e = env or os.environ
    return str(e.get("JWT_SECRET") or e.get("CLAWHIVE_JWT_SECRET") or "").strip()


def is_agent_browser_auth_enabled(env: dict[str, str] | None = None) -> bool:
    e = env or os.environ
    raw = str(e.get("AGENT_BROWSER_AUTH") or e.get("MANAGER_USER_AUTH") or "").strip().lower()
    if raw in ("0", "false", "no"):
        return False
    if raw in ("1", "true", "yes"):
        return True
    return bool(resolve_jwt_secret(e))


def _b64url_decode(data: str) -> bytes:
    s = data.replace("-", "+").replace("_", "/")
    pad = "=" * (-len(s) % 4)
    return base64.b64decode(s + pad)


def extract_bearer(authorization: str | None) -> str:
    raw = str(authorization or "").strip()
    if raw.lower().startswith("bearer "):
        return raw[7:].strip()
    return ""


def verify_clawhive_jwt(token: str, env: dict[str, str] | None = None) -> dict[str, Any]:
    secret = resolve_jwt_secret(env)
    if not secret:
        raise ClawhiveAuthError("JWT_SECRET_not_configured")
    parts = str(token or "").strip().split(".")
    if len(parts) != 3:
        raise ClawhiveAuthError("jwt_malformed")
    header_b64, payload_b64, sig_b64 = parts
    signing_input = f"{header_b64}.{payload_b64}".encode("utf-8")
    expected = hmac.new(secret.encode("utf-8"), signing_input, hashlib.sha256).digest()
    try:
        got = _b64url_decode(sig_b64)
    except Exception as exc:  # noqa: BLE001
        raise ClawhiveAuthError("jwt_sig_invalid") from exc
    if not hmac.compare_digest(got, expected):
        raise ClawhiveAuthError("jwt_sig_invalid")
    try:
        header = json.loads(_b64url_decode(header_b64))
        payload = json.loads(_b64url_decode(payload_b64))
    except Exception as exc:  # noqa: BLE001
        raise ClawhiveAuthError("jwt_malformed") from exc
    if str(header.get("alg") or "HS256") != "HS256":
        raise ClawhiveAuthError("jwt_alg_unsupported")
    sub = str(payload.get("sub") or "").strip()
    if not sub:
        raise ClawhiveAuthError("jwt_missing_sub")
    exp = payload.get("exp")
    if exp is not None:
        try:
            if time.time() >= float(exp):
                raise ClawhiveAuthError("jwt_expired")
        except ClawhiveAuthError:
            raise
        except Exception as exc:  # noqa: BLE001
            raise ClawhiveAuthError("jwt_malformed") from exc
    return {
        "sub": sub,
        "role": str(payload.get("role") or "viewer"),
        "tenant_id": str(payload.get("tenant_id") or "default"),
        "user_id": sub,
        "username": sub,
    }


def resolve_internal_token(headers: dict[str, str], env: dict[str, str] | None = None) -> bool:
    e = env or os.environ
    expected = str(e.get("CLAWHIVE_INTERNAL_TOKEN") or e.get("AGENT_INTERNAL_TOKEN") or "").strip()
    if not expected:
        return False
    got = (
        str(headers.get("x-clawhive-internal-token") or headers.get("X-Clawhive-Internal-Token") or "").strip()
        or str(headers.get("x-internal-token") or headers.get("X-Internal-Token") or "").strip()
    )
    return bool(got and hmac.compare_digest(got, expected))


def require_browser_or_internal(
    headers: dict[str, str],
    *,
    env: dict[str, str] | None = None,
    authorization: str | None = None,
) -> dict[str, Any]:
    """
    返回 {mode: internal|browser|open, user?: dict}
    失败抛 ClawhiveAuthError(code=login_required|invalid_user_token|...)
    """
    e = env or os.environ
    hdrs = {str(k).lower(): str(v) for k, v in (headers or {}).items()}
    if resolve_internal_token(hdrs, e):
        return {"mode": "internal"}
    if not is_agent_browser_auth_enabled(e):
        return {"mode": "open"}
    token = extract_bearer(authorization or hdrs.get("authorization"))
    if not token:
        token = str(hdrs.get("x-clawhive-user-token") or "").strip()
    if not token:
        raise ClawhiveAuthError("login_required")
    try:
        user = verify_clawhive_jwt(token, e)
    except ClawhiveAuthError:
        raise
    except Exception as exc:  # noqa: BLE001
        raise ClawhiveAuthError("invalid_user_token") from exc
    return {"mode": "browser", "user": user}
